indico/modules/events/registration/badges.py

```diff
@@ -65,11 +65,17 @@ def _build_pdf(self, canvas):
         for person, (x, y) in zip(self.persons, self._iter_position(canvas, n_horizontal, n_vertical)):
             self._draw_badge(canvas, person, self.template, self.tpl_data, x * cm, y * cm)
 
+    def _remove_invisible_items(self, tpl_data):
+        pos_to_remove = [i for i, item in enumerate(tpl_data.items) if not tpl_data.layers[item['layer']]]
+        for pos in reversed(pos_to_remove):
+            tpl_data.items.pop(pos)
```

Comment:
I think this is dangerous: AFAICT `items` is passed through directly from the dict that's on the model, so the only reason why this does not affect the stored data is that there's no mutation detection on mutable sqlalchemy data.

I think this is a safer alternative (untested):

```py
tpl_data = tpl_data._replace(items=[item for item in tpl_data.items if tpl_data.layers[item['layer']]])
```

Reply:
Oops, indeed. However `_replace` creates a new instance and `self.tpl_data` will still refer to the old one. (side note: passing instance attributes to `_draw_badge` method seems unnecessary)
Actually it would be more reasonable to remove all invisible items only once right at the beginning. However that would break our intention to manipulate item's visibility later per registrations.
So eventually I wouldn't do any destructive modification only skipping invisible items when placing them on the canvas.